src/fsharp/TypeChecker.fs

```diff
@@ -8594,37 +8594,87 @@ and delayRest rest mPrior delayed =
         let mPriorAndLongId = unionRanges mPrior (rangeOfLid longId)
         DelayedDotLookup (rest, mPriorAndLongId) :: delayed
 
+/// Typecheck "nameof" expressions
+and TcNameOfExpr cenv env tpenv (synArg: SynExpr) = 
+
+    let rec stripParens expr =
+        match expr with
+        | SynExpr.Paren(expr, _, _, _) -> stripParens expr
+        | _ -> expr
+
+    let cleanSynArg = stripParens synArg
+    let m = cleanSynArg.Range
+    let rec check overallTyOpt expr (delayed: DelayedItem list) = 
+        match expr with
+        | LongOrSingleIdent (false, (LongIdentWithDots(longId, _) as lidd), _, _) when longId.Length > 0 ->
+            let ad = env.eAccessRights
+            let id, rest = List.headAndTail longId
+            match ResolveLongIndentAsModuleOrNamespace cenv.tcSink ResultCollectionSettings.AllResults cenv.amap m true OpenQualified env.eNameResEnv ad id rest true with 
+            | Result modref when delayed.IsEmpty && modref |> List.exists (p23 >> IsEntityAccessible cenv.amap m ad) -> 
+                () // resolved to a module or namespace, done with checks
+            | _ -> 
+                let (TypeNameResolutionInfo(_, staticArgsInfo)) = GetLongIdentTypeNameInfo delayed
+                match ResolveTypeLongIdent cenv.tcSink cenv.nameResolver ItemOccurence.UseInAttribute OpenQualified env.eNameResEnv ad longId staticArgsInfo PermitDirectReferenceToGeneratedType.No with
+                | Result tcref when IsEntityAccessible cenv.amap m ad tcref -> 
+                    () // resolved to a type name, done with checks
+                | _ -> 
+                    let overallTy = match overallTyOpt with None -> NewInferenceType() | Some t -> t 
+                    let _, _ = TcLongIdentThen cenv overallTy env tpenv lidd delayed
```

Comment:
We are doing this check for its side effect. The risk is that a future change would add a new argument to the function, giving
```
    let f x y = ....
    ...
    f x |> ignore
```
i.e. a curried function is created and then thrown away.  One way in F# programming to avoid this is
```
    let (* some annotation or type requirement *) = f x
```
e.g.
```
    let _, _ = f x
```
or
```
    let _ : result-type = f x
```
or
```
    f x |> ignore<result-type>
```
so that when the `y` argument gets added you get an error.

In this case, requiring the tuple return was the simplest way to lay down a constraint.

Reply:
Interesting.  Thanks for the explanation.

Reply:
I think this advice should be added the F# Coding Conventions

Reply:
What about a warning if `ignore` is used on a value of `FSharpFunc` type?

Reply:
An ignore on an `FSharpFunc` that is partially applied could work, as it wouldn't warn on perfectly valid `f x |> ignore` calls where `f` has one parameter.

Reply:
Yes, I would be in favour of such a warning being added, on by default (subject to `/langversion:5.0` switch if necessary)